File: app/core/configRegistry.ts
import { ICustomerConfig } from '../interfaces/config/customerConfig.interface';
import { oneSpaConfig } from '../../../customers/configurations/one_spa/oneSpaConfig';
import { oneBarnConfig } from '../../../customers/configurations/one_barn/oneBarnConfig';

// Configuration Registry Interface
export interface IConfigRegistry {
  getCustomerConfig(customerId: string): ICustomerConfig | null;
  getAllCustomerIds(): string[];
  isValidCustomer(customerId: string): boolean;
  getCustomerByDomain(domain: string): ICustomerConfig | null;
  getActiveCustomers(): ICustomerConfig[];
  validateConfig(config: ICustomerConfig): IConfigValidationResult;
}

export interface IConfigValidationResult {
  isValid: boolean;
  errors: string[];
  warnings: string[];
}

// Configuration Registry Implementation
class ConfigRegistry implements IConfigRegistry {
  private readonly configs: Map<string, ICustomerConfig> = new Map();

  constructor() {
    this.loadConfigurations();
  }

  private loadConfigurations(): void {
    // Register all customer configurations
    this.registerConfig(oneSpaConfig);
    this.registerConfig(oneBarnConfig);
    
    // Future customer configs would be registered here:
    // this.registerConfig(financial_services_config);
    // this.registerConfig(property_management_config);
    
    console.log(`Loaded ${this.configs.size} customer configurations`);
  }

  private registerConfig(config: ICustomerConfig): void {
    const validation = this.validateConfig(config);
    
    if (!validation.isValid) {
      throw new Error(`Invalid configuration for ${config.customerId}: ${validation.errors.join(', ')}`);
    }

    if (validation.warnings.length > 0) {
      console.warn(`Configuration warnings for ${config.customerId}:`, validation.warnings);
    }

    this.configs.set(config.customerId, config);
    console.log(`Registered configuration for: ${config.customerName} (${config.customerId})`);
  }

  public getCustomerConfig(customerId: string): ICustomerConfig | null {
    return this.configs.get(customerId) || null;
  }

  public getAllCustomerIds(): string[] {
    return Array.from(this.configs.keys());
  }

  public isValidCustomer(customerId: string): boolean {
    return this.configs.has(customerId);
  }

  public getCustomerByDomain(domain: string): ICustomerConfig | null {
    for (const config of this.configs.values()) {
      // Check main branding domain
      if (config.branding && domain.includes(config.branding.companyName.toLowerCase().replace(/\s+/g, ''))) {
        return config;
      }

      // Check tenant domains
      for (const tenant of config.tenants) {
        if (tenant.domain === domain || domain.includes(tenant.subdomain)) {
          return config;
        }
      }
    }
    return null;
  }

  public getActiveCustomers(): ICustomerConfig[] {
    return Array.from(this.configs.values()).filter(config => 
      config.tenants.some(tenant => tenant.isActive)
    );
  }

  public validateConfig(config: ICustomerConfig): IConfigValidationResult {
    const errors: string[] = [];
    const warnings: string[] = [];

    // Required field validations
    if (!config.customerId) {
      errors.push('customerId is required');
    }

    if (!config.customerName) {
      errors.push('customerName is required');
    }

    if (!config.industryType) {
      errors.push('industryType is required');
    }

    // Branding validations
    if (!config.branding?.companyName) {
      errors.push('branding.companyName is required');
    }

    if (!config.branding?.colors?.primary) {
      errors.push('branding.colors.primary is required');
    }

    // Location validations
    if (!config.locations || config.locations.length === 0) {
      errors.push('At least one location is required');
    } else {
      config.locations.forEach((location, index) => {
        if (!location.id) {
          errors.push(`Location ${index} missing id`);
        }
        if (!location.name) {
          errors.push(`Location ${index} missing name`);
        }
        if (!location.address?.street) {
          errors.push(`Location ${index} missing address.street`);
        }
      });
    }

    // Tenant validations
    if (!config.tenants || config.tenants.length === 0) {
      errors.push('At least one tenant is required');
    } else {
      config.tenants.forEach((tenant, index) => {
        if (!tenant.id) {
          errors.push(`Tenant ${index} missing id`);
        }
        if (!tenant.domain) {
          errors.push(`Tenant ${index} missing domain`);
        }
      });
    }

    // Security validations
    if (config.security?.authentication?.passwordPolicy?.minLength < 8) {
      warnings.push('Password minimum length should be at least 8 characters');
    }

    if (!config.security?.authentication?.mfaRequired && config.compliance?.hipaa?.enabled) {
      warnings.push('MFA should be required for HIPAA compliance');
    }

    // Compliance validations
    if (config.compliance?.hipaa?.enabled && !config.security?.encryption?.dataAtRest) {
      errors.push('Data at rest encryption is required for HIPAA compliance');
    }

    if (config.compliance?.pciDss?.enabled && !config.security?.encryption?.dataInTransit) {
      errors.push('Data in transit encryption is required for PCI DSS compliance');
    }

    return {
      isValid: errors.length === 0,
      errors,
      warnings
    };
  }
}

// Singleton instance
export const configRegistry = new ConfigRegistry();

// Helper functions following the established pattern
export const getCustomerConfig = (customerId: string): ICustomerConfig | null => {
  return configRegistry.getCustomerConfig(customerId);
};

export const getCustomerByDomain = (domain: string): ICustomerConfig | null => {
  return configRegistry.getCustomerByDomain(domain);
};

export const getAllCustomerIds = (): string[] => {
  return configRegistry.getAllCustomerIds();
};

export const isValidCustomer = (customerId: string): boolean => {
  return configRegistry.isValidCustomer(customerId);
};

export const getActiveCustomers = (): ICustomerConfig[] => {
  return configRegistry.getActiveCustomers();
};

export const validateCustomerConfig = (config: ICustomerConfig): IConfigValidationResult => {
  return configRegistry.validateConfig(config);
};

// Configuration utilities
export const getCustomerEnvironment = (customerId: string): string => {
  const config = getCustomerConfig(customerId);
  return config?.environment || 'development';
};

export const getCustomerComplianceFrameworks = (customerId: string): string[] => {
  const config = getCustomerConfig(customerId);
  if (!config) return [];

  return Object.entries(config.compliance)
    .filter(([key, value]) => 
      typeof value === 'object' && 
      value !== null && 
      'enabled' in value && 
      value.enabled
    )
    .map(([key]) => key.toUpperCase());
};

export const getCustomerBrandingVars = (customerId: string): Record<string, string> => {
  const config = getCustomerConfig(customerId);
  if (!config?.branding) return {};

  const { colors, fonts } = config.branding;
  return {
    '--primary-color': colors.primary,
    '--secondary-color': colors.secondary,
    '--accent-color': colors.accent || '#F5F5F5',
    '--background-color': colors.background,
    '--text-color': colors.text,
    '--primary-font': fonts.primary,
    '--secondary-font': fonts.secondary || fonts.primary,
  };
};

export const getCustomerTenantBySubdomain = (customerId: string, subdomain: string): any => {
  const config = getCustomerConfig(customerId);
  if (!config) return null;

  return config.tenants.find(tenant => tenant.subdomain === subdomain);
};

export const getCustomerIntegrationConfig = (customerId: string, integrationName: string): any => {
  const config = getCustomerConfig(customerId);
  if (!config) return null;

  return config.integrations.thirdParty[integrationName] || null;
};

// Export for use in FastAPI backend
export default configRegistry; 

File: app/core/configRegistry.py
"""
Configuration Registry for OneVault Platform
==========================================

Python wrapper around TypeScript configuration system providing
async interface for FastAPI integration.
"""

import asyncio
import json
import subprocess
from pathlib import Path
from typing import Dict, Any, List, Optional
from datetime import datetime, time
import logging

from .config import settings, TypeScriptConfigBridge
from ..config.configConstants import (
    CONFIG_FIELDS, 
    API_KEYS, 
    CSS_VARS,
    get_nested_config_value,
    get_branding_css_vars,
    validate_required_config_fields
)

logger = logging.getLogger(__name__)

class ConfigRegistry:
    """
    Async configuration registry that interfaces with TypeScript configs.
    Provides caching, validation, and helper functions for the FastAPI application.
    """
    
    def __init__(self):
        self.ts_bridge = TypeScriptConfigBridge(settings)
        self._initialized = False
        self._platform_config_cache: Optional[Dict[str, Any]] = None
        self._customer_cache: Dict[str, Dict[str, Any]] = {}
        self._cache_timestamp = datetime.utcnow()
        
    async def initialize(self) -> None:
        """Initialize the configuration registry."""
        try:
            # Test TypeScript bridge connectivity
            test_result = await self._execute_async_ts_function("getAllCustomerIds()")
            if test_result is not None:
                self._initialized = True
                logger.info("✅ ConfigRegistry initialized successfully")
            else:
                raise Exception("Failed to connect to TypeScript configuration system")
        except Exception as e:
            logger.error(f"❌ ConfigRegistry initialization failed: {e}")
            raise
    
    async def _execute_async_ts_function(self, function_call: str) -> Any:
        """Execute TypeScript function asynchronously."""
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            None, 
            self.ts_bridge._execute_typescript_function, 
            function_call
        )
    
    async def get_platform_config(self) -> Dict[str, Any]:
        """Get platform-wide configuration."""
        if self._platform_config_cache is None:
            # Load platform configuration from TypeScript
            platform_config = await self._execute_async_ts_function("getPlatformConfig()")
            
            if platform_config is None:
                # Fallback to default platform config
                platform_config = {
                    "supportedIndustries": [
                        "spa_wellness",
                        "financial_services", 
                        "equestrian",
                        "property_management",
                        "healthcare"
                    ],
                    "complianceFrameworks": [
                        "hipaa", "hitech", "sox", "pci_dss", 
                        "gdpr", "ccpa", "glba"
                    ],
                    "features": {
                        "multiTenant": True,
                        "whiteLabel": True,
                        "industryModules": True,
                        "dataVault2": True,
                        "auditLogging": True,
                        "encryptionAtRest": True
                    },
                    "pricing": {
                        "baseMonthly": 4999,
                        "perLocation": 299,
                        "enterpriseDiscount": 0.15
                    }
                }
            
            self._platform_config_cache = platform_config
            
        return self._platform_config_cache
    
    async def get_customer_config(self, customer_id: str) -> Optional[Dict[str, Any]]:
        """Get customer configuration with caching."""
        cache_key = f"customer_{customer_id}"
        
        # Check cache first
        if cache_key in self._customer_cache:
            cached_config = self._customer_cache[cache_key]
            # Check if cache is still valid (5 minutes)
            if (datetime.utcnow() - self._cache_timestamp).seconds < 300:
                return cached_config
        
        # Load from TypeScript config
        config = await self._execute_async_ts_function(f"getCustomerConfig('{customer_id}')")
        
        if config:
            self._customer_cache[cache_key] = config
            self._cache_timestamp = datetime.utcnow()
        
        return config
    
    async def get_all_customer_ids(self) -> List[str]:
        """Get all registered customer IDs."""
        result = await self._execute_async_ts_function("getAllCustomerIds()")
        return result if result else []
    
    async def get_customer_by_domain(self, domain: str) -> Optional[Dict[str, Any]]:
        """Get customer configuration by domain."""
        return await self._execute_async_ts_function(f"getCustomerByDomain('{domain}')")
    
    async def validate_customer_config(self, customer_id: str) -> Dict[str, Any]:
        """Validate customer configuration."""
        try:
            config = await self.get_customer_config(customer_id)
            
            if not config:
                return {
                    API_KEYS.VALIDATION_VALID: False,
                    API_KEYS.VALIDATION_ERRORS: [f"Customer configuration not found: {customer_id}"],
                    API_KEYS.VALIDATION_WARNINGS: []
                }
            
            errors = []
            warnings = []
            
            # Validate required fields using constants
            is_valid, missing_fields = validate_required_config_fields(config)
            if not is_valid:
                errors.extend([f"Missing required field: {field}" for field in missing_fields])
            
            # Validate customer info
            customer_info = get_nested_config_value(config, CONFIG_FIELDS.CUSTOMER, default={})
            if customer_info:
                if not customer_info.get(CONFIG_FIELDS.CUSTOMER_NAME):
                    errors.append("Customer name is required")
                if not customer_info.get(CONFIG_FIELDS.CUSTOMER_EMAIL):
                    warnings.append("Customer contact email not specified")
            
            # Validate locations
            locations = get_nested_config_value(config, CONFIG_FIELDS.LOCATIONS, default=[])
            if not locations:
                warnings.append("No locations configured")
            else:
                for i, location in enumerate(locations):
                    if not location.get(CONFIG_FIELDS.LOCATION_NAME):
                        errors.append(f"Location {i+1} missing name")
                    if not location.get(CONFIG_FIELDS.LOCATION_ADDRESS):
                        warnings.append(f"Location {i+1} missing address")
            
            # Validate pricing
            pricing = get_nested_config_value(config, CONFIG_FIELDS.PRICING, default={})
            if pricing and not pricing.get(CONFIG_FIELDS.PRICING_MONTHLY_TOTAL):
                errors.append("Monthly total pricing not calculated")
            
            return {
                API_KEYS.VALIDATION_VALID: len(errors) == 0,
                API_KEYS.VALIDATION_ERRORS: errors,
                API_KEYS.VALIDATION_WARNINGS: warnings,
                API_KEYS.CONFIG_SUMMARY: {
                    API_KEYS.CUSTOMER_NAME: get_nested_config_value(
                        config, CONFIG_FIELDS.CUSTOMER, CONFIG_FIELDS.CUSTOMER_NAME
                    ),
                    CONFIG_FIELDS.INDUSTRY: get_nested_config_value(config, CONFIG_FIELDS.INDUSTRY),
                    API_KEYS.LOCATION_COUNT: len(locations),
                    API_KEYS.MONTHLY_COST: get_nested_config_value(
                        config, CONFIG_FIELDS.PRICING, CONFIG_FIELDS.PRICING_MONTHLY_TOTAL, default=0
                    )
                }
            }
            
        except Exception as e:
            logger.error(f"Config validation error for {customer_id}: {e}")
            return {
                "valid": False,
                "errors": [f"Validation error: {str(e)}"],
                "warnings": []
            }
    
    async def get_customer_branding_css(self, customer_id: str) -> Dict[str, str]:
        """Get customer branding as CSS variables."""
        config = await self.get_customer_config(customer_id)
        
        branding = get_nested_config_value(config, CONFIG_FIELDS.BRANDING, default={})
        if not branding:
            return {}
        
        return get_branding_css_vars(branding)
    
    async def is_location_open_now(self, customer_id: str, location_id: str) -> bool:
        """Check if a specific location is currently open."""
        config = await self.get_customer_config(customer_id)
        
        if not config or "locations" not in config:
            return False
        
        # Find the location
        location = None
        for loc in config["locations"]:
            if loc.get("id") == location_id:
                location = loc
                break
        
        if not location or not location.get("isActive", True):
            return False
        
        # Check business hours
        business_hours = location.get("businessHours")
        if not business_hours:
            return True  # Assume open if no hours specified
        
        now = datetime.now()
        current_day = now.strftime("%A").lower()
        current_time = now.time()
        
        day_hours = business_hours.get(current_day)
        if not day_hours:
            return False  # Closed if no hours for today
        
        if day_hours.get("closed", False):
            return False
        
        try:
            open_time = time.fromisoformat(day_hours.get("open", "00:00"))
            close_time = time.fromisoformat(day_hours.get("close", "23:59"))
            
            return open_time <= current_time <= close_time
        except (ValueError, TypeError):
            return True  # Default to open if time parsing fails
    
    async def get_customer_compliance_frameworks(self, customer_id: str) -> List[str]:
        """Get compliance frameworks for customer."""
        config = await self.get_customer_config(customer_id)
        
        if not config or "compliance" not in config:
            return []
        
        return config["compliance"].get("frameworks", [])
    
    async def get_customer_locations_summary(self, customer_id: str) -> Dict[str, Any]:
        """Get summary of customer locations."""
        config = await self.get_customer_config(customer_id)
        
        if not config or "locations" not in config:
            return {
                "total": 0,
                "active": 0,
                "open_now": 0,
                "locations": []
            }
        
        locations = config["locations"]
        active_count = 0
        open_now_count = 0
        
        for location in locations:
            if location.get("isActive", True):
                active_count += 1
                
                if await self.is_location_open_now(customer_id, location.get("id")):
                    open_now_count += 1
        
        return {
            "total": len(locations),
            "active": active_count,
            "open_now": open_now_count,
            "locations": [
                {
                    "id": loc.get("id"),
                    "name": loc.get("name"),
                    "isActive": loc.get("isActive", True),
                    "isOpenNow": await self.is_location_open_now(customer_id, loc.get("id"))
                }
                for loc in locations
            ]
        }
    
    async def reload_all_configs(self) -> None:
        """Reload all configurations from TypeScript files."""
        try:
            # Clear caches
            self._platform_config_cache = None
            self._customer_cache.clear()
            self.ts_bridge.clear_cache()
            
            # Reinitialize
            await self.initialize()
            
            logger.info("✅ All configurations reloaded successfully")
            
        except Exception as e:
            logger.error(f"❌ Configuration reload failed: {e}")
            raise
    
    async def get_customer_pricing_summary(self, customer_id: str) -> Dict[str, Any]:
        """Get customer pricing summary."""
        config = await self.get_customer_config(customer_id)
        
        if not config or "pricing" not in config:
            return {
                "monthlyTotal": 0,
                "baseCost": 0,
                "locationCost": 0,
                "locationCount": 0,
                "annualTotal": 0
            }
        
        pricing = config["pricing"]
        location_count = len(config.get("locations", []))
        monthly_total = pricing.get("monthlyTotal", 0)
        
        return {
            "monthlyTotal": monthly_total,
            "baseCost": pricing.get("baseCost", 0),
            "locationCost": pricing.get("locationCost", 0),
            "locationCount": location_count,
            "annualTotal": monthly_total * 12,
            "currency": "USD"
        }
    
    def is_initialized(self) -> bool:
        """Check if registry is initialized."""
        return self._initialized 